### ISSUE TITLE
Subscribing to connectionLost outside default MqttCallback

### ISSUE THREAD

**Issue by pabsilva** [eclipse-paho/paho.mqtt.java#620]
Hi,

I was wondering: is it possible to subscribe to the "connectionLost" event without setting the default message listener (i.e. mqttClient.setCallback(new MqttCallback() ...)?

I am asking because I subscribing to the topics using the "subscribe(topic, qos, IMqttMessageListener)" function. I noticed that, having set CleanSession=false, all pending qos>0 messages will be delivered right after being connected (to the default MqttCallback), but before I have the chance to call the subscribe functions (where I have assigned the handling functions). I also cannot subscribe before the connection is established, otherwise exceptions will be thrown because the "client is not connected". I am not sure if this is an expected behavior, so I was wondering how to be able to use IMqttMessageListener subscriptions and yet still be notified when a connection is lost.

**Comment by spyro2000** [comment]
That's actually a very good question, I've basically the same problem. When I set cleanSesssion to FALSE, I just get 'socket errors' when trying to connect with QoS > 0.

**Comment by miketran78727** [comment]
Since MQTT 3.1.0, CONNECT packet must be the first packet: see http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718028 so this Paho client will throw exception if the internal state is not connected (We don't want to violate MQTT protocol).

The `lostConnection()` callback is intended to notify the application that the network connect has been lost. The server could close the connect because of error or there is a network issue on the client side.

If you use MqttClient, the connect() method will wait for completion, then you can use `isConnected()` to make sure the client is connected before calling subscribe()

If you use AsyncMqttClient, the connect() is asynchronous.  You may want to wait for completion ( `connect(options).waitForCompletion(timeout)` )  and use `isConnected()` to check on connection status. Or you use MqttCallbackExtended callback https://github.com/eclipse/paho.mqtt.java/blob/master/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/MqttCallbackExtended.java#L32 to wait connection status.

**Comment by pabsilva** [comment]
Yes, I was using MqttClient and I tried the MqttCallbackExtended and that's the where the problem begins: I can't subscribe to lostConnection() without, at the time time, set the default message listener (which I don't want to, because it causes the problems described before).

**Comment by icraggs** [comment]
Ok, I think this is the same issue as #595.  We need a way of setting the individual subscription handler callbacks before the connection is made.
